File: mobile/android-shell/app/src/main/java/com/codexapp/mobile/NavigationPolicy.kt
package com.codexapp.mobile

import android.net.Uri

enum class NavigationDecision {
    InApp,
    External,
    Blocked,
}

object NavigationPolicy {
    fun evaluate(uri: Uri?): NavigationDecision {
        if (uri == null) {
            return NavigationDecision.Blocked
        }

        val scheme = (uri.scheme ?: "").lowercase()
        val host = (uri.host ?: "").lowercase()

        if (scheme in ShellConfig.externalSchemes) {
            return NavigationDecision.External
        }
        if (scheme !in ShellConfig.allowedSchemes) {
            return NavigationDecision.Blocked
        }
        if (host in ShellConfig.allowedHosts) {
            return NavigationDecision.InApp
        }
        if (host in ShellConfig.externalHosts) {
            return NavigationDecision.External
        }
        return NavigationDecision.Blocked
    }
}

File: mobile/ios-shell/CodexApp/NavigationPolicy.swift
import Foundation

enum NavigationDecision {
    case inApp
    case external
    case blocked
}

enum NavigationPolicy {
    static func evaluate(_ url: URL?) -> NavigationDecision {
        guard let url else {
            return .blocked
        }

        let scheme = (url.scheme ?? "").lowercased()
        let host = (url.host ?? "").lowercased()

        if ShellConfig.externalSchemes.contains(scheme) {
            return .external
        }
        guard ShellConfig.allowedSchemes.contains(scheme) else {
            return .blocked
        }
        if ShellConfig.allowedHosts.contains(host) {
            return .inApp
        }
        if ShellConfig.externalHosts.contains(host) {
            return .external
        }
        return .blocked
    }
}
